sdk/tests/conformance2/textures/misc/tex-unpack-params.html

```diff
@@ -0,0 +1,409 @@
+                  gl.RGB, gl.UNSIGNED_BYTE, array);
+  }
+  wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "buffer too small");
+
+  // Uncomment the next line to make the test pass on NVIDIA Linux drivers.
+  // They incorrectly require padding for the last row.
+  // bufferSize += imageSizes.padding;
+  array = setupArrayBuffer(bufferSize, true);
+  if (useUnpackBuffer) {
+    gl.bufferData(gl.PIXEL_UNPACK_BUFFER, array, gl.DYNAMIC_DRAW);
+  }
+  gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGB8, testCase.width, testCase.height, testCase.depth, 0,
+                gl.RGB, gl.UNSIGNED_BYTE, useUnpackBuffer ? 0 : array);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "texImage3D with correct buffer size should succeed");
+
+  // TODO(zmo): Verify texture data.
+
+  debug("TexSubImage3D : xoffset = " + testCase.xoffset + ", yoffset = " + testCase.yoffset +
+        ", zoffset = " + testCase.zoffset);
+  var subWidth = testCase.width - testCase.xoffset;
+  var subHeight = testCase.height - testCase.yoffset;
+  var subDepth = testCase.depth - testCase.zoffset;
+  imageSizes = computeImageSizes3D(subWidth, subHeight, subDepth, testCase);
+  bufferSize = imageSizes.totalSize;
+
+  if (useUnpackBuffer) {
+    gl.bufferData(gl.PIXEL_UNPACK_BUFFER, bufferSize - 1, gl.DYNAMIC_DRAW);
+    gl.texSubImage3D(gl.TEXTURE_3D, 0, testCase.xoffset, testCase.yoffset, testCase.zoffset,
+                     subWidth, subHeight, subDepth, gl.RGB, gl.UNSIGNED_BYTE, 0);
+  } else {
+    array = setupArrayBuffer(bufferSize - 1, false);
+    gl.texSubImage3D(gl.TEXTURE_3D, 0, testCase.xoffset, testCase.yoffset, testCase.zoffset,
+                     subWidth, subHeight, subDepth, gl.RGB, gl.UNSIGNED_BYTE, array);
+  }
+  wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "buffer too small");
+
+  // Uncomment the next line to make the test pass on NVIDIA Linux drivers.
+  // They incorrectly require padding for the last row.
+  // bufferSize += imageSizes.padding;
+  array = setupArrayBuffer(bufferSize, true);
+  if (useUnpackBuffer) {
+    gl.bufferData(gl.PIXEL_UNPACK_BUFFER, array, gl.DYNAMIC_DRAW);
+  }
+  gl.texSubImage3D(gl.TEXTURE_3D, 0, testCase.xoffset, testCase.yoffset, testCase.zoffset,
+                   subWidth, subHeight, subDepth,
+                   gl.RGB, gl.UNSIGNED_BYTE, useUnpackBuffer ? 0 : array);
+  wtu.glErrorShouldBe(gl, gl.NO_ERROR, "texSubImage3D with correct buffer size should succeed");
+
+  // TODO(zmo): Verify texture data
+
+  if (buffer) {
+    gl.bindBuffer(gl.PIXEL_UNPACK_BUFFER, null);
+    gl.deleteBuffer(buffer);
+  }
+  gl.bindTexture(gl.TEXTURE_3D, null);
+  gl.deleteTexture(tex);
+}
+
+function runTests() {
+  var gl = wtu.create3DContext("example");
+  if (!gl) {
+    testFailed("Fail to get a WebGL context");
+    return;
+  }
+
+  // For 2D cases, depth, zoffset, imageHeight, skipImages are ignored.
+  var testCases = [
+    { width: 5, height: 7, depth: 3, xoffset: 2, yoffset: 3, zoffset: 1,
```

Comment:
This is getting a bit pedantic, but it would be nice to see some tests with different width/height/depth/offsets, just to make sure we don't happen to be hitting some really weird edge case. Specifically it would be good to do a few tests where the textures are powers of 2. That should be the "simple" case, but it's worth testing.

Also, if these are going to be the same for a large number of cases in may be better to patch them in down below rather than repeat them over and over. For example:

```
for (var ii = 0; ii < testCases.length; ++ii) {
    var testCase = testCases[ii];
    testCase.width = 5;
    testCase.height = 7;
    // Etc.
    runTestIteration2D(gl, testCase, useUnpackBuffer);
    runTestIteration3D(gl, testCase, useUnpackBuffer);

    testCase.width = 4;
    testCase.height = 8;
    // Etc.
    runTestIteration2D(gl, testCase, useUnpackBuffer);
    runTestIteration3D(gl, testCase, useUnpackBuffer);
 }
```